catalog: Allow for more builtin roles

jkosh44 opened the pull request:
This commit updates role ids to have two separate namespaces. One for
builtin roles and one for user roles. This will allow us to more easily
add builtin roles in the future.

This is a breaking change.


### Motivation
This PR refactors existing code.


### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Changes role ids to be a string, which changes the `id` column of `mz_roles`.

mjibson commented on test/cluster/mzcompose.py:
```diff
@@ -297,10 +297,11 @@ def workflow_test_builtin_migration(c: Composition) -> None:
-            Materialized(options=mz_options)
+            # This commit added the missing column rolconnlimit to pg_authid.
+            Materialized(
+                image="materialize/materialized:devel-438ea318093b3a15a924fbdae70e0db6d379a921",
```
I'm real sad if this needs to happen, because it effectively means this test will never test any code in a PR, but just run exactly the same code over and over.

This test has caught migration bugs from other PRs before, and seems kinda good to have maybe? At least, I'd argue that if we want to hard code this we should maybe just...remove the test because it's doing nothing and replace it with a real migration test.

jkosh44 replied:
I deleted it, it had a good run but with a breaking change it will no longer work. I'll open up an issue tonight about making adding a new upgrade test for this feature.

mjibson commented on src/adapter/src/catalog/storage.rs:
```diff
@@ -1601,8 +1609,8 @@ impl_codec!(ItemValue);
 struct RoleKey {
-    #[prost(uint64)]
-    id: u64,
+    #[prost(string)]
```
Oooh our first protobuf migration. The code here will cause any existing stash to fail on startup. We need to decide if we're ok with breaking the world again or not. I don't know.

jkosh44 replied:
As per slack discussion, next release will have another breaking change, so we're going to piggy back off of that.

jkosh44 commented:
Closing for #14145

benesch commented:
Skimmed the test changes and I love it. Deferring the actual review to @mjibson though.

jkosh44 commented:
@mjibson In case you already started reviewing, I just changed the name of the added user from "mz_system_external" to "http_default_user" because I thought the old name was really non-descriptive.

jkosh44 commented:
> @mjibson In case you already started reviewing, I just changed the name of the added user from "mz_system_external" to "http_default_user" because I thought the old name was really non-descriptive.

Oops didn't follow my own rule and gave it a bad name without a "mz_" prefix. Just fixed the name of the role.

mjibson commented on src/adapter/src/catalog/builtin.rs:
```diff
@@ -2288,11 +2293,13 @@ pub static BUILTINS_STATIC: Lazy<Vec<Builtin<NameReference>>> = Lazy::new(|| {
 });
-pub static BUILTIN_ROLES: Lazy<Vec<BuiltinRole>> = Lazy::new(|| vec![MZ_SYSTEM]);
+pub static BUILTIN_ROLES: Lazy<Vec<BuiltinRole>> =
+    Lazy::new(|| vec![MZ_SYSTEM, MZ_HTTP_DEFAULT_USER]);
```
Is it required that the order of this doesn't change should we add more? That is, new entries must be added at the end instead of being alphabetically sorted? If so, document that with a comment.

jkosh44 replied:
The order for this doesn't matter. It's only used on startup to persist any new roles that haven't already been persisted.

mjibson commented on src/adapter/src/catalog/storage.rs:
```diff
@@ -112,12 +114,18 @@ async fn migrate<S: Append>(
             txn.id_allocator.insert(
                 IdAllocKey {
-                    name: ROLE_ID_ALLOC_KEY.into(),
+                    name: USER_ROLE_ID_ALLOC_KEY.into(),
```
If this is a breaking change to which we aren't providing an upgrade path, the cloud team should be flagged.

jkosh44 replied:
https://materializeinc.slack.com/archives/CL68GT3AT/p1660169050996079

mjibson commented on test/cluster/mzcompose.py:
```diff
@@ -234,86 +232,3 @@ def workflow_test_drop_default_cluster(c: Composition) -> None:
     c.sql("CREATE CLUSTER default REPLICAS (default (SIZE '1'))")
-
-
-def workflow_test_builtin_migration(c: Composition) -> None:
```
rip

mjibson commented on src/adapter/src/catalog/builtin.rs:
```diff
@@ -2101,9 +2105,10 @@ AS SELECT
-    id: 0,
+pub const MZ_SYSTEM: BuiltinRole = BuiltinRole { name: "mz_system" };
+
+pub const MZ_HTTP_DEFAULT_USER: BuiltinRole = BuiltinRole {
```
Is this unused? Should it be added in a later PR?

jkosh44 replied:
I thought it might be good to test a second builtin role, but you're probably right. I should save this for a later PR.
